### PR TITLE
 libvirt: Add Terraform variables for memory/CPU, bump master to 4GiB 

### PR DESCRIPTION
**Pull request by cgwalters**
My main dev environment is a Lenovo P50 with 64GB of RAM - I got
it specifically to run some large VMs (and/or many VMs) specifically
with OpenShift in mind.

Increasing RAM on my master to 8GB is a *very* noticeable speed improvement
and I think reliabilty; before I saw the apiserver be `OOMKilled`
sometimes, and `kswapd0` was constantly doing writeback.

These variables aren't bubbled all the way up to the documented
installer config, but one can now do e.g.:

```
$ env TF_VAR_tectonic_libvirt_memory=8192 TF_VAR_tectonic_libvirt_vcpu=4 ./bin/openshift-install create cluster --dir=osiris
```

Previously:
    
  - https://github.com/openshift/installer/pull/408
  - https://github.com/openshift/installer/pull/163

### CODE REVIEW

**Comment by ashcrow** [comment]
This is great :heart: 

One ask though would be to have a follow on to document it in user docs.

**Comment by rphillips** [comment]
Thoughts on bumping the default to 4096?

**Comment by cgwalters** [comment]
> Thoughts on bumping the default to 4096?

TL;DR: yes

Longer 2¢ - libvirt is primarily a developer target; we should maintain a "functional baseline" for it.  I can imagine people wanting to use it on a "real" server to kick the tires a bit more seriously before e.g. switching to using OpenStack or bare metal.

For developers, 16GB laptops are common I believe.  If we target the one master, one worker case, with 4GiB each + 2GiB for bootstrap, we're using almost 2/3% of the machine's resources.  This isn't a problem for *me* but we may need to think about what we can do to make it easier for people who just need a basic cluster to test something.

Now, one issue is that right now workers and masters get the same amount of RAM.  That seems definitely worth fixing.

**Comment by cgwalters** [comment]
Updated to default to 4GiB.

**Comment by wking** [comment]
> (I shouldn't approve installer changes directly IMHO)

`/lgtm` is for any org member.  Obviously you can make your own decisions about when to use it, but I have no problem with anyone dropping `/lgtm` if they want to.  Some more on the expected `/lgtm` vs. `/approve` semantics over [here][1].

[1]: https://github.com/kubernetes/test-infra/pull/9138#issuecomment-431168278

**Comment by abhinavdahiya** [comment]
Is there is a specific process that is using up all of that memory. I would expect a little bit more investigation before we bump the RAM again.

_firefox/chrome can eat up quite a bit of GBs so not much room left :P_

**Comment by wking** [comment]
> Is there is a specific process that is using up all of that memory. I would expect a little bit more investigation before we bump the RAM again.

I've seen folks pointing fingers at Prometheus, but haven't measured myself.  @crawford may be able to clarify.

And while I'm in favor of trimming memory back, I like stability too ;).  Unfortunately, it's hard to get time allocated for "claw back memory", e.g. see @sjenning promising to [get us back under 2GB once we dropped `kube-core-operators`][1] (which happened in #420). :p

[1]: https://github.com/openshift/installer/pull/408#issuecomment-426742377

**Comment by cgwalters** [comment]
> firefox/chrome can eat up quite a bit of GBs so not much room left :P

This is an interesting topic actually - I don't have a good link to hand but at least a few years ago Firefox basically defaulted to using at least ⅓ of available RAM or so.  More recently came along "tab idling" where tabs you haven't used in a while get basically torn down except for their URL.  Part of implementing this is ensuring that the browser serializes TextArea data (like the one I'm typing in now!) so you don't lose data.

This is a bit like [pod idling](https://docs.openshift.com/container-platform/3.6/admin_guide/idling_applications.html) - but without CRIU your app needs to be storing its state on a PV (or be stateless).

But there's a latency tradeoff here - which is why browsers use RAM.  (Also, it's not browsers that use RAM mostly, it's websites)

File: data/data/libvirt/variables-libvirt.tf [hunk 1]
```diff
@@ -32,3 +32,15 @@ variable "tectonic_libvirt_master_ips" {
   type        = "list"
   description = "the list of desired master ips. Must match tectonic_master_count"
 }
+
+variable "tectonic_libvirt_memory" {
```
[the remark this answers is not in the document]
Comment (cgwalters) [reply data/data/libvirt/variables-libvirt.tf]: `master_memory` and not `libvirt_master_memory`?

**Comment by cgwalters** [comment]
Updated :new:

File: data/data/libvirt/variables-libvirt.tf (hunk 1 above)
[the remark this answers is not in the document]
Comment (wking) [reply data/data/libvirt/variables-libvirt.tf]: We're not going to launch a libvirt-and-AWS-at-the-same-time cluster, so I don't see a need for platforms in Terraform variable names.  But we certainly have plenty of existing names like that, so I'm fine with you going that way if you like.

File: data/data/libvirt/variables-libvirt.tf [hunk 2]
```diff
@@ -32,3 +32,19 @@ variable "tectonic_libvirt_master_ips" {
   type        = "list"
   description = "the list of desired master ips. Must match tectonic_master_count"
 }
+
+# It's definitely recommended to bump this if you can.
+variable "libvirt_master_memory" {
+  type        = "string"
+  description = "RAM in MiB allocated to master and nodes"
```
Comment (wking) [inline data/data/libvirt/variables-libvirt.tf]: This is just for master.

File: data/data/libvirt/variables-libvirt.tf [hunk 3]
```diff
@@ -32,3 +32,19 @@ variable "tectonic_libvirt_master_ips" {
...
+  default     = "4096"
+}
+
+# At some point this one is likely to default to the number
+# of physical cores you have.  See also
+# https://pagure.io/standard-test-roles/pull-request/223
+variable "libvirt_master_vcpu" {
+  type        = "string"
+  description = "CPUs allocated to master and nodes"
```
Comment (wking) [inline data/data/libvirt/variables-libvirt.tf]: Also here, this is just for masters.

**Comment by cgwalters** [comment]
Updated again to fix descriptions.

**Comment by wking** [comment]
/lgtm

On the, ah, wildly optimistic side, tunables could be used by folks trying to push memory back down too ;).

**Comment by crawford** [comment]
Mirroring @wking's lgtm. It looks like the bot missed the message.

/lgtm

**Comment by richm** [comment]
@openshift/sig-logging

**Comment by wking** [comment]
It's just documented in this PR at the moment, but the coming origin rebase should help with both memory and CPU and the planned CPU-reservation adjustments will help with CPU.
